flang/lib/Lower/Bridge.cpp

```diff
@@ -1444,26 +1472,27 @@ class FirConverter : public Fortran::lower::AbstractConverter {
   // gen expression, if any; share code with END of procedure
   void genFIR(Fortran::lower::pft::Evaluation &eval,
               const Fortran::parser::ReturnStmt &stmt) {
-    const auto *funit = eval.getOwningProcedure();
-    assert(funit && "not inside main program or a procedure");
+    auto *funit = eval.getOwningProcedure();
+    assert(funit && "not inside main program or a subprogram");
     if (funit->isMainProgram()) {
       genExitRoutine();
-    } else {
-      if (stmt.v) {
-        // Alternate return
-        TODO();
-      }
-      // an ordinary RETURN should be lowered as a GOTO to the last block of the
-      // SUBROUTINE
-      auto *subr = eval.getOwningProcedure();
-      assert(subr && "RETURN not in a PROCEDURE");
-      if (!subr->finalBlock) {
-        auto insPt = builder->saveInsertionPoint();
-        subr->finalBlock = builder->createBlock(&builder->getRegion());
-        builder->restoreInsertionPoint(insPt);
-      }
-      builder->create<mlir::BranchOp>(toLocation(), subr->finalBlock);
+      return;
     }
+    if (stmt.v) {
+      // Alternate return statement -- assign alternate return index.
+      const auto altReturnIndex = builder->createIntegerConstant(
+          builder->getIndexType(),
+          *Fortran::evaluate::ToInt64(*Fortran::semantics::GetExpr(*stmt.v)));
```

Comment:
This code assumes that alternate return index  is a constant (you cannot call `Fortran::evaluate::ToInt64` on something else than a integer constant, it would return nullptr.)
I believe alternate return is a _scalar-int-expr_ that does not have to be a constant.
I think you should rather have:
```
auto expr = Fortran::semantics::GetExpr(*stmt.v);
assert(expr && "missing typed expr")
auto altReturnIndex  = builder->createHere<fir::ConvertOp>(builder->getIndexType(), genExprValue(*expr));
```

BTW the `const` is not wrong but on `mlir::Value` that is a small C++ type passed through copy rather than reference, it is not really needed. Better to add the const if in doubt though.